// feathers-socketio/lib/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

exports.default = function (options, config) {
  if (typeof options === 'function') {
    config = options;
    options = {};
  }

  return function () {
    var app = this;

    app.configure(_feathersSocketCommons2.default);

    _uberproto2.default.mixin({
      setup: function setup(server) {
        var io = this.io = _socket2.default.listen(server, options);

        io.use(function (socket, next) {
          socket.feathers = { provider: 'socketio' };
          next();
        });

        if (typeof config === 'function') {
          debug('Calling SocketIO configuration function');
          config.call(this, io);
        }

        this._socketInfo = {
          method: 'emit',
          connection: function connection() {
            return io.sockets;
          },
          clients: function clients() {
            return io.sockets.sockets;
          },
          params: function params(socket) {
            return socket.feathers;
          }
        };

        return this._super.apply(this, arguments);
      }
    }, app);
  };
};

var _debug = require('debug');

var _debug2 = _interopRequireDefault(_debug);

var _socket = require('socket.io');

var _socket2 = _interopRequireDefault(_socket);

var _uberproto = require('uberproto');

var _uberproto2 = _interopRequireDefault(_uberproto);

var _feathersSocketCommons = require('feathers-socket-commons');

var _feathersSocketCommons2 = _interopRequireDefault(_feathersSocketCommons);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var debug = (0, _debug2.default)('feathers-socketio');

module.exports = exports['default'];
